Asset basenames in Jest snapshots

myrjola opened the pull request:
This change only affects tests run with Jest. `require('/images/image1.png')` will be replaced with

```
Object {
  "testUri": "relative/path/to/images/image1.png",
}
``` 

in the Jest snapshot instead of always being 1 returned by RelativeImageStub. This change makes it possible to test conditional asset loading in components.

The problem with this change is that it will probably break a lot of existing snapshots, but that should be easily fixed when a project updates to a new version of React Native by running `jest -u` to update all snapshots.

## Motivation:

A component can have conditional asset loading based on its props, this logic would be nice to test with Jest snapshots. This problem has been discussed in https://github.com/facebook/jest/issues/2838.

## Test Plan:

I have added a snapshot test `assetRelativePathInSnapshot`.

```
$ npm test -- assetRelativePathInSnapshot

> react-native@1000.0.0 test /Users/martin/git/react-native
> jest "assetRelativePathInSnapshot"

 PASS  Libraries/Image/__tests__/assetBaseNameInSnapshot.js
  ✓ renders assets based on basename (132ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   1 passed, 1 total
Time:        1.674s
Ran all test suites matching "assetBasenameInSnapshot".
```

It is easy to verify from the snapshot that the change works.

Libraries/Image/__tests__/__snapshots__/assetBaseNameInSnapshot.js.snap:
```
// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders assets based on relative path 1`] = `
<View>
  <Image
    source={
      Object {
        "testUri": "Libraries/Image/__tests__/img/img1.png",
      }
    }
  />
  <Image
    source={
      Object {
        "testUri": "Libraries/Image/__tests__/img/img2.png",
      }
    }
  />
</View>
`;
```

Without the change the output would look like this:

```
// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders assets based on basename 1`] = `
<View>
  <Image
    source={1}
  />
  <Image
    source={1}
  />
</View>
`;
```

## Breaking changes:

### Jest snapshots will need to be updated

* **Who does this affect**: Everyone using `Image` in Jest snapshots
* **How to migrate**: Running `jest -u` will update the snapshots, the snapshots should be reviewed that they are correct.
* **Why make this breaking change**: It enables testing of conditional asset loading.
* **Severity (number of people affected x effort)**: Low.

ide commented on jest/assetFileTransformer.js:
```diff
@@ -0,0 +1,19 @@
+  // the correct images are loaded for components. So require('img1.png')
+  // becomes 'img1.png' in the Jest snapshot.
+  process: (_, filename) =>
+    `module.exports = ${JSON.stringify(path.basename(filename))};`
```
Better to mimic what the actual transformer does and return an object.

myrjola commented:
Rebased the code and addressed the requested changes. I also chose to return relative path instead of basename. Updated the PR description with the latest changes and also added a breaking changes section.

myrjola commented on jest/assetFileTransformer.js:
```diff
@@ -8,12 +8,18 @@
-    `module.exports = ${JSON.stringify(path.basename(filename))};`
+    `module.exports = {uri: ${JSON.stringify(path.relative('.', filename))}};`,
+
+  getCacheKey: createCacheKeyFunction([__filename]),
```
This function was needed after rebasing on latest master. I took some inspiration from https://github.com/facebook/react-native/blob/master/jest/preprocessor.js.

Without it running Jest will give the following error:

```
 FAIL  Libraries/Image/__tests__/assetBaseNameInSnapshot.js
  ● renders assets based on basename

    /Users/martin/git/react-native/Libraries/Image/__tests__/img/img1.png:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,global,jest){module.exports = [object Object];
                                                                                                                      ^^^^^^

    SyntaxError: Unexpected identifier

      at ScriptTransformer._transformAndBuildScript (node_modules/jest-runtime/build/script_transformer.js:305:17)
      at Object.<anonymous> (Libraries/Image/__tests__/assetBaseNameInSnapshot.js:21:64)
      at tryCallTwo (node_modules/promise/lib/core.js:45:5)
      at doResolve (node_modules/promise/lib/core.js:200:13)

  ✕ renders assets based on basename (144ms)
```

ide commented on jest/assetFileTransformer.js:
```diff
@@ -0,0 +1,25 @@
+  // require('img1.png') becomes `Object { "uri": 'path/to/img1.png' }` in the
+  // Jest snapshot.
+  process: (_, filename) =>
+    `module.exports = {uri: ${JSON.stringify(path.relative('.', filename))}};`,
```
This "uri" field doesn't actually exist at runtime, maybe call it testUri so it's clear this field is defined only in tests?

ide commented:
@cpojer how does this look to you?

myrjola commented:
Renamed uri to testUri in the latest commit and updated description.

cpojer commented:
@mjesun mind taking care of this one?

mjesun commented:
Sorry for the late reply; we haven't forgot about that, but we had some issues importing the PR :( (unrelated to the PR itself).

mjesun commented:
@myrjola Just a small heads up; [we adjusted the test](https://github.com/facebook/react-native/commit/33bbfb7125569d3f3b26f030f7db4a7e8cdeabf4) so it is able to run on our infra (and potentially anywhere regardless of the exec folder 🙂. Thanks again for the contribution! 😊

oblador commented on jest-preset.json:
```diff
@@ -7,12 +7,15 @@
     ]
   },
   "moduleNameMapper": {
-    "^[./a-zA-Z0-9$_-]+\\.(bmp|gif|jpg|jpeg|png|psd|svg|webp)$": "RelativeImageStub",
```
Removing this broke importing files with multiple device resolution variations. Metro bundler will resolve `require('./file.png')` to `./file@2x.png` if you're on a 2x screen.

myrjola replied:
Thanks for reporting! Could you give a bit more detail what the problem is? Apparently this change has broken your tests. Could you give some steps to reproduce?

I immediately notice that the regex `^[./a-zA-Z0-9$_-]+\\.(bmp|gif|jpg|jpeg|png|psd|svg|webp)$` does not match the `@` character in `./file@2x.png`. The regex has been moved from `moduleNameMapper` to the `transform` mapper on line 17 in the same file. Perhaps the change from `./file.png` to `./file@2x.png` happens between the mappers. @oblador, could you edit the regex on line 17 to `^[./a-zA-Z0-9@$_-]+\\.(bmp|gif|jpg|jpeg|png|psd|svg|webp)$`, so that it also matches `./file@2x.png`?

sam-vdp replied:
This also breaks tests completely on Windows for me. Please see #19370 for details
